File: source/table/hand.c
#include "table/table.h"

static int rankValue(Table_Rank rank) {
    if(rank == TABLE_RANK_ACE) {
        return 11;
    }

    if(rank >= TABLE_RANK_TEN) {
        return 10;
    }

    return (int)rank + 1;
}

int Table_handValue(Table_Hand hand) {
    int value = 0;
    int aceCount = 0;

    for(int index = 0; index < hand.count; index += 1) {
        value += rankValue(hand.cards[index].rank);

        if(hand.cards[index].rank == TABLE_RANK_ACE) {
            aceCount += 1;
        }
    }

    while(value > 21 && aceCount > 0) {
        value -= 10;
        aceCount -= 1;
    }

    return value;
}

bool Table_handIsSoft(Table_Hand hand) {
    int value = 0;
    int aceCount = 0;

    for(int index = 0; index < hand.count; index += 1) {
        value += rankValue(hand.cards[index].rank);

        if(hand.cards[index].rank == TABLE_RANK_ACE) {
            aceCount += 1;
        }
    }

    while(value > 21 && aceCount > 0) {
        value -= 10;
        aceCount -= 1;
    }

    return aceCount > 0;
}

File: tools/strategy/source/simulation/table/hand.py
"""Hand primitives mirroring the C table module."""

from dataclasses import dataclass, field

from simulation.table.card import Card, Rank, rank_value


@dataclass
class Hand:
    """Blackjack hand."""

    cards: list[Card] = field(default_factory=list)


def value(hand: Hand) -> int:
    """Return the best blackjack value for a hand."""
    total = 0
    ace_count = 0

    for card in hand.cards:
        total += rank_value(card.rank)

        if card.rank == Rank.ACE:
            ace_count += 1

    while total > 21 and ace_count > 0:
        total -= 10
        ace_count -= 1

    return total


def is_soft(hand: Hand) -> bool:
    """Return true when the hand contains an ace counted as eleven."""
    total = 0
    ace_count = 0

    for card in hand.cards:
        total += rank_value(card.rank)

        if card.rank == Rank.ACE:
            ace_count += 1

    while total > 21 and ace_count > 0:
        total -= 10
        ace_count -= 1

    return ace_count > 0
